WalletWasabi.Fluent/ViewModels/Wallets/Receive/ReceiveAddressViewModel.cs

```diff
@@ -1,64 +1,70 @@
 using System.Threading.Tasks;
 using System.Windows.Input;
-using Avalonia;
-using Gma.QrCodeNet.Encoding;
-using NBitcoin;
+using DynamicData;
 using ReactiveUI;
-using WalletWasabi.Blockchain.Analysis.Clustering;
-using WalletWasabi.Blockchain.Keys;
 using WalletWasabi.Fluent.Extensions;
+using WalletWasabi.Fluent.Models.UI;
+using WalletWasabi.Fluent.Models.Wallets;
 using WalletWasabi.Fluent.ViewModels.Navigation;
-using WalletWasabi.Hwi;
 using WalletWasabi.Logging;
-using WalletWasabi.Wallets;
 
 namespace WalletWasabi.Fluent.ViewModels.Wallets.Receive;
 
 [NavigationMetaData(Title = "Receive Address")]
 public partial class ReceiveAddressViewModel : RoutableViewModel
 {
-	private readonly Wallet _wallet;
-	private readonly HdPubKey _model;
+	private readonly ObservableAsPropertyHelper<bool[,]> _qrCode;
 
-	public ReceiveAddressViewModel(Wallet wallet, HdPubKey model)
+	public ReceiveAddressViewModel(UiContext uiContext, IWalletModel wallet, IAddress model, bool isAutoCopyEnabled)
 	{
-		_wallet = wallet;
-		_model = model;
-		Address = _model.GetP2wpkhAddress(_wallet.Network).ToString();
-		Labels = _model.Label;
-		IsHardwareWallet = _wallet.KeyManager.IsHardwareWallet;
-		IsAutoCopyEnabled = Services.UiConfig.Autocopy;
-
-		GenerateQrCode();
+		UiContext = uiContext;
+			Model = model;
+		Address = model.Text;
+		Labels = model.Labels;
+		IsHardwareWallet = wallet.IsHardwareWallet();
 
 		SetupCancel(enableCancel: false, enableCancelOnEscape: true, enableCancelOnPressed: true);
 
 		EnableBack = true;
 
-		CopyAddressCommand = ReactiveCommand.CreateFromTask(async () =>
-		{
-			if (Application.Current is { Clipboard: { } clipboard })
-			{
-				await clipboard.SetTextAsync(Address);
-			}
-		});
+		CopyAddressCommand = ReactiveCommand.CreateFromTask(() => UiContext.Clipboard.SetTextAsync(Address));
 
-		ShowOnHwWalletCommand = ReactiveCommand.CreateFromTask(async () => await OnShowOnHwWalletAsync(_model, _wallet.Network, _wallet.KeyManager.MasterFingerprint));
+		ShowOnHwWalletCommand = ReactiveCommand.CreateFromTask(ShowOnHwWalletAsync);
 
-		SaveQrCodeCommand = ReactiveCommand.CreateFromTask(async () => await OnSaveQrCodeAsync());
+		SaveQrCodeCommand = ReactiveCommand.CreateFromTask(OnSaveQrCodeAsync);
 
 		SaveQrCodeCommand.ThrownExceptions
 			.ObserveOn(RxApp.TaskpoolScheduler)
 			.Subscribe(ex => Logger.LogError(ex));
 
 		NextCommand = CancelCommand;
+
+		GenerateQrCodeCommand = ReactiveCommand.CreateFromObservable(() => UiContext.QrCodeGenerator.Generate(model.Text));
+		_qrCode = GenerateQrCodeCommand.ToProperty(this, nameof(QrCode));
+
+		if (isAutoCopyEnabled)
+		{
+			CopyAddressCommand.Execute(null);
+		}
+
+		GenerateQrCodeCommand
+			.Execute()
+			.Subscribe();
+
+		wallet.Addresses
+			.Watch(model.Text)
+			.Where(change => change.Current.IsUsed)
+			.Do(_ => UiContext.Navigate(NavigationTarget.Default).Back())
+			.Subscribe();
 	}
 
+	public ReactiveCommand<Unit, bool[,]> GenerateQrCodeCommand { get; }
```

Comment:
Things getting messy regarding the commands. We agreed to always use `ICommand` for consistency.

(it seems like it is already a mistake `QrCodeCommand`, but that should be fixed too)